### PR TITLE
Add sargX alias to support Go function arguments

### LINKED ISSUE #740: Support golang arguments
using gccgo, Go function arguments can be read using arg0...argN, as it follows the standard AMD64 ABI. Eg:

```
# bpftrace -e 'uprobe:/home/bgregg/func:main.add { printf("%d %d\n", arg0, arg1); }'
Attaching 1 probe...
42 13
```

But using Go gc, it follows the Plan9 stack convention where everything is placed on the stack. The same arguments can be read instead using:

```
# bpftrace -e 'uprobe:/home/bgregg/Lang/go/func:main*add { printf("%d %d\n", *(reg("sp") + 8), *(reg("sp") + 16)); }'
Attaching 1 probe...
42 13
```

This ticket is to suggest adding these aliases:

- `goarg0` == `*(reg("sp") + 8)`
- `goarg1` == `*(reg("sp") + 16)`
- etc

Maybe better: call them `sarg0`, `sarg1`, ...: short for "stack argument", since Golang may not be the only runtime/compiler doing this.

### PR DESCRIPTION
**Pull request by acj**
~I played with two different implementations and finally settled on this lexer spec approach. Adding something to the parser was another possibility, but it was fairly ugly and referenced a lot more types.~

~If this lexer rule sees an `sarg[0-9]` token, it swallows the token and then uses lex's `unput` to feed the stack expression back into the input stream (in reverse order). It's then lexed and parsed as normal.~

Resolves #740.

### CODE REVIEW

File: tests/runtime/builtin [hunk 1]
```diff
@@ -58,6 +58,11 @@ RUN bpftrace -v -e 'k:vfs_read { printf("SUCCESS '$test' %d\n", arg0); exit(); }
 EXPECT SUCCESS arg -?[0-9][0-9]*
 TIMEOUT 5
 
+NAME sarg
+RUN bpftrace -v -e 'k:vfs_read { printf("SUCCESS '$test' %d\n", sarg0); exit(); }'
+EXPECT SUCCESS sarg -?[0-9][0-9]*
```
Comment (ajor) [inline tests/runtime/builtin]: What's the value this will be printing?

**Comment by acj** [comment]
Thanks for the feedback. I'll give that a try.

**Comment by brendangregg** [comment]
sarg is good. That's how I documented it in Chapter 12.

> A future version of bpftrace may support these as aliases, eg, as sarg0,
> sarg1 [BIBLIOG186], short for "stack argument".

:-)

**Comment by acj** [comment]
Okay, I've updated this to do its work during codegen. It seems to work well for functions that take integers. Both `argX` and `sargX` behave incorrectly for Go functions that take a mix of strings and ints (or other things) regardless of compiler; I'll open a separate issue for that.

For test coverage, is it acceptable to add a small Go binary built with `gc` along with its source? Other options?

Hasn't been tested with USDT yet.

File: src/ast/codegen_llvm.cpp [hunk 2]
```diff
@@ -177,6 +177,38 @@ void CodegenLLVM::visit(Builtin &builtin)
...
+    if (sp_offset == -1)
+    {
+      std::cerr << "negative offset for stack pointer" << std::endl;
+      abort();
+    }
+
+    AllocaInst *dst;
+    Value *src;
+
+    dst = b_.CreateAllocaBPF(SizedType(Type::integer, 8), "reg_sp");
```
Comment (fbs) [inline src/ast/codegen_llvm.cpp]: You should be able to do this without a proberead, see 7d0aa993433dd8b743898c8e1743d72b5af75684 which was merged last night.

File: README.md [hunk 3]
```diff
@@ -257,6 +257,7 @@ Variables:
 - `stack` - Kernel stack trace
 - `ustack` - User stack trace
 - `arg0`, `arg1`, ... etc. - Arguments to the function being traced
+- `sarg0`, `sarg1`, ... etc. - Arguments to the function being traced (for Go programs built with `gc`)
```
Comment (fbs) [inline README.md]: The go part is a bit confusing here, unless you know what `go gc` is. I would reword it to make it clear that it is reading from the stack.

File: src/ast/semantic_analyser.cpp [hunk 4]
```diff
@@ -162,6 +162,22 @@ void SemanticAnalyser::visit(Builtin &builtin)
       err_ << arch::name() << " doesn't support " << builtin.ident << std::endl;
     builtin.type = SizedType(Type::integer, 8);
   }
+  else if (!builtin.ident.compare(0, 4, "sarg") && builtin.ident.size() == 5 &&
```
Comment (fbs) [inline src/ast/semantic_analyser.cpp]: This is basically identical to the "args" one right, maybe those can be merged to avoid duplicate code?

Comment (acj) [reply src/ast/semantic_analyser.cpp]: After the most recent changes, these two else-if blocks have diverged a bit.

**Comment by acj** [comment]
Thanks, all! I've integrated your feedback and added tests.

> Another option to avoid a Go dependency could be to compile a regular C program with a function that takes more arguments than x86_64 has registers for

I think that's a good idea and will provide good enough coverage for now.

**Comment by acj** [comment]
@ajor This is ready for another look.

**Comment by acj** [comment]
Rebased onto master.

@fbs @ajor Any other feedback? Can this be merged?

File: tests/testprogs/stack_args.c [hunk 5]
```diff
@@ -0,0 +1,8 @@
+// Declare enough args that some are placed on the stack
+void too_many_args(int a, int b, int c, int d, int e, int f, int g, int h) {}
```
Comment (fbs) [inline tests/testprogs/stack_args.c]: This adds a lot of warning to the build output as all args are unused, something like this will fix it:
```
#define ARG(x) int (x) __attribute((unused))

void too_many_args(ARG( a), ARG( b), ARG( c), ARG( d), ARG( e), ARG( f), ARG( g), ARG( h)) {}

int main(void)
```

Comment (acj) [reply tests/testprogs/stack_args.c]: Oof, yes it does. Nice catch.

**Comment by danobi** [comment]
Some small things:
* Could you rebase onto master so CI can run again?
* @olsajiri , do you think semantic analyser should reject `sarg*` usage if uprobe is used with virtual address / function+offset? B/c you'd have no idea what the stack is other than at function entry
  * I'd probably lean towards allowing it b/c you're already in "I know what I'm doing territory"

Other than that, LGTM

**Comment by acj** [comment]
@danobi Done.

BTW, I had to install the `binutils-dev` package to build current master on Ubuntu 19.10. I was getting this error, which was misleading and took a bit of searching to track down:

```
-- Please install the libbpf development package (missing: LIBBPF_LIBRARIES LIBBPF_INCLUDE_DIRS)
-- Please install the libbfd development package (missing: LIBBFD_LIBRARIES LIBBFD_INCLUDE_DIRS)
-- Performing Test LIBBFD_DISASM_FOUR_ARGS_SIGNATURE
-- Performing Test LIBBFD_DISASM_FOUR_ARGS_SIGNATURE - Failed
-- Please install the libopcodes development package (missing: LIBOPCODES_LIBRARIES LIBOPCODES_INCLUDE_DIRS)
-- Looking for sys/sdt.h
-- Looking for sys/sdt.h - not found
-- Looking for pthread.h
-- Looking for pthread.h - not found
CMake Error at /usr/share/cmake-3.13/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find Threads (missing: Threads_FOUND)
Call Stack (most recent call first):
  /usr/share/cmake-3.13/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.13/Modules/FindThreads.cmake:205 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  tests/CMakeLists.txt:88 (find_package)
```

I think this was a recent change because I was able to build master on the same machine a few days ago.

**Comment by mmarchini** [comment]
> do you think semantic analyser should reject sarg* usage if uprobe is used with virtual address / function+offset? B/c you'd have no idea what the stack is other than at function entry

I think that makes sense, maybe allowing that only on unsafe mode. Or at least output a warning by default suggesting users to use a probe w/o offset tô capture the arguments (which can be stored in a map and used in the probe w/ offset later)

**Comment by fbs** [comment]
> > do you think semantic analyser should reject sarg* usage if uprobe is used with virtual address / function+offset? B/c you'd have no idea what the stack is other than at function entry
> 
> I think that makes sense, maybe allowing that only on unsafe mode. Or at least output a warning by default suggesting users to use a probe w/o offset tô capture the arguments (which can be stored in a map and used in the probe w/ offset later)

+1 for a warning. I don't think it's unsafe, I don't see how it could crash your program/system. But in most cases it will probably behave unexpected so a warning would be good.

**Comment by acj** [comment]
Sounds good. I added a warning:

```
$ bpftrace -e 'uprobe:./tests/testprogs/stack_args:0x1148 { printf("%d\n", sarg0); }'
stdin:1:61-66: WARNING: Using an address offset with the sargX built-in can lead to unexpected behavior. Consider using a probe without an offset.
uprobe:./tests/testprogs/stack_args:0x1148 { printf("%d\n", sarg0); }
                                                            ~~~~~
Attaching 1 probe...
^C

$ bpftrace -e 'uprobe:./tests/testprogs/stack_args:too_many_args+24 { printf("%d\n", sarg0); }'
stdin:1:71-76: WARNING: Using an address offset with the sargX built-in can lead to unexpected behavior. Consider using a probe without an offset.
uprobe:./tests/testprogs/stack_args:too_many_args+24 { printf("%d\n", sarg0); }
                                                                      ~~~~~
Attaching 1 probe...
^C
```

File: src/ast/semantic_analyser.cpp [hunk 6]
```diff
@@ -170,6 +170,26 @@ void SemanticAnalyser::visit(Builtin &builtin)
...
+      if (type != ProbeType::kprobe && type != ProbeType::uprobe)
+        err_ << "The " << builtin.ident << " builtin can only be used with "
+             << "'kprobes' and 'uprobes' probes" << std::endl;
+
+      if (is_final_pass() &&
+          (attach_point->address != 0 || attach_point->func_offset != 0)) {
+        std::stringstream warn_buf;
+        warn_buf << "Using an address offset with the sargX built-in can lead "
+                 << "to unexpected behavior. Consider using a probe without an "
+                 << "offset.";
```
Comment (fbs) [inline src/ast/semantic_analyser.cpp]: I think just having the first line is enough. A probe without offset isn't really an alternative you can use if you need offsets. The alternative you do have is to store `sargs` in a map on function entry and use that in the offset probe. But I think we're  better off documenting that somewhat instead of trying to put it all in the warning message.

File: src/ast/semantic_analyser.cpp [hunk 7]
```diff
@@ -170,6 +170,27 @@ void SemanticAnalyser::visit(Builtin &builtin)
...
+        err_ << "The " << builtin.ident << " builtin can only be used with "
+             << "'kprobes' and 'uprobes' probes" << std::endl;
+
+      if (is_final_pass() &&
+          (attach_point->address != 0 || attach_point->func_offset != 0)) {
+        // If sargX values are needed when using an offset, they can be stored in a map
+        // when entering the function and then referenced from an offset-based probe
+        std::stringstream warn_buf;
+        warn_buf << "Using an address offset with the sargX built-in can lead to "
+                 << "unexpected behavior";
```
Comment (acj) [inline src/ast/semantic_analyser.cpp]: @mmarchini @fbs Highlighting this for any final wordsmithing

**Comment by fbs** [comment]
Nope looks good to me. 

Thanks @acj !
